### ISSUE TITLE
Handling jar dependency resolution with jruby projects

### ISSUE THREAD

**Issue by talevy** [elastic/logstash#5514]
since there are more and more java-jar based ruby plugins being introduced, jar conflicts are inevitable. To remedy this, one strategy we can try and practice is to package the java extension using jar-dependencies. This way we do not have farjar/shadowed-jars, and we can rely on jar-dependencies' `require_jar` to attempt to error out on version conflicts.

here is an example project using this build strategy: https://github.com/talevy/log4j2-json-logging

it is a joint effort between the `Rakefile`, `build.gradle`, and the gemspec.

This affects how Logstash can support multiple version of the same dependencies that get pulled into its classpath. Currently, `jar_dependencies` requires these jars and checks if any other jars of the same type were loaded previously. Here is its logic: https://github.com/mkristian/jar-dependencies/blob/master/lib/jar_dependencies.rb#L302-L320.

Moving forward, we need to decide how to organize these projects. the proposal above is just one way.

**Comment by guyboertje** [comment]
@talevy - I was hoping you could elaborate on the sequence of actors and actions, i.e. ...
T0: developer/jarvis runs `gem build`, this 1) compiles Java and publishes it to a local `.m2` folder; 2) jar_dependencies vendors the jar from the local Maven repo.
T2: the gem is published to rubygems with the vendored jar(s) and NO local Maven repo and NO java source.
T1: developer/jarvis/user runs `bundle install` or equivalent and jar_dependencies will resolve jars across all jruby gems.

Please correct me - I am sure I have missed something.

**Comment by talevy** [comment]
ah.

bootstrapping dependencies:
```
$ bundle install
```

testing

```
$ rake test
```

packaging gem

```
$ rake package
```

cleaning up build directories and files

```
$ rake clean
```

**Comment by guyboertje** [comment]
@talevy, @colinsurprenant, @andrewvc 
How fussy are we on including the java source in the gem and not the jar?
I wonder whether we can move the compile, publish and vendor into a gem post_install hook?
```ruby
Gem.post_install do |gem_installer|
  compile_publish_to_local_maven_repo
  vendor_jars
end
```
See [jar_dependencies post_install hook](https://github.com/mkristian/jar-dependencies/blob/master/lib/jars/post_install_hook.rb)

Will the user be left high and dry when it does not compile with the Java the user has?
Will this work in Elastic Cloud?

Here I am not talking about development on the java gem itself but that all of our java gems that Logstash depends on will compile and vendor jars when we do `bin/bundle install` or `bin/logstash-plugin install`.

**Comment by talevy** [comment]
I would prefer this myself, but the reason we have traditionally vendored jars when building the gem because it makes supporting offline plugin installation more straightforward to support

**Comment by guyboertje** [comment]
Fair point.

**Comment by purbon** [comment]
@talevy yes, and no. with bundler package at my understanding jars will be also included for offline management, so offline is covered also.

**Comment by ph** [comment]
@guyboertje I used the strategy proposed by @talevy  in https://github.com/logstash-plugins/logstash-input-beats/pull/93/ I did not have any hard issues or problem with it.
